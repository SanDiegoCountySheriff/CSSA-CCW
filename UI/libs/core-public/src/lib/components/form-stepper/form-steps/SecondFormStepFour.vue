<template>
  <div>
    <v-container>
      <v-row class="mb-5">
        <v-col
          cols="12"
          lg="6"
        >
          <v-form
            ref="form"
            v-model="state.valid"
          >
            <v-text-field
              :label="$t('Signature')"
              :rules="[v => !!v || $t(' Signature cannot be blank ')]"
              v-model="state.signature"
            />
          </v-form>
        </v-col>
        <v-col
          cols="12"
          lg="6"
        >
          <div class="signature-preview">
            <canvas
              ref="signatureCanvas"
              height="100"
              width="250"
            ></canvas>
            <v-btn
              text
              class="m-5"
              color="error"
              @click="handleCanvasClear"
            >
              {{ $t('clear') }}
            </v-btn>
          </div>
        </v-col>
      </v-row>
      <v-divider class="mb-5" />
      <FormButtonContainer
        :valid="state.valid"
        @submit="handleSubmit"
        @save="router.push('/')"
        @cancel="router.push('/')"
        @back="handlePreviousSection"
      />
    </v-container>
    <v-snackbar
      :value="state.snackbar"
      :timeout="3000"
      bottom
      color="error"
      outlined
    >
      {{ $t('File upload unsuccessful please try again.') }}
    </v-snackbar>
  </div>
</template>

<script setup lang="ts">
import FormButtonContainer from '@core-public/components/containers/FormButtonContainer.vue';
import Routes from '@core-public/router/routes';
import axios from 'axios';
import { useCompleteApplicationStore } from '@core-public/stores/completeApplication';
import { useMutation } from '@tanstack/vue-query';
import { useRouter } from 'vue-router/composables';
import { reactive, ref, watch } from 'vue';

interface ISecondFormStepFourProps {
  handleNextSection: CallableFunction;
  handlePreviousSection: CallableFunction;
}

const props = defineProps<ISecondFormStepFourProps>();
const signatureCanvas = ref<HTMLCanvasElement | null>(null);
const applicationStore = useCompleteApplicationStore();
const router = useRouter();

const state = reactive({
  valid: false,
  file: {},
  signature: '',
  snackbar: false,
});

const fileMutation = useMutation({
  mutationFn: handleFileUpload,
  onSuccess: () => {
    applicationStore.completeApplication.application.currentStep = 9;
    router.push(Routes.QUALIFYING_QUESTIONS_ROUTE_PATH);
  },
  onError: () => {
    state.snackbar = true;
  },
});

async function handleSubmit() {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //@ts-ignore
  const image = signatureCanvas.value.toDataURL('image/jpeg', 0.5);
  const file = new File([image], 'file');
  const form = new FormData();

  form.append('fileToPersist', file);

  state.file = form;

  fileMutation.mutate();
}

async function handleFileUpload() {
  const newFileName = `${applicationStore.completeApplication.application.orderId}_${applicationStore.completeApplication.application.personalInfo.lastName}_${applicationStore.completeApplication.application.personalInfo.firstName}_signature`;

  await axios.post(
    `http://localhost:5148/Api/Document/v1/Document/uploadApplicantFile?saveAsFileName=${newFileName}`,
    state.file
  );
}

function handleCanvasClear() {
  const canvas = signatureCanvas.value;
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //@ts-ignore
  const ctx = canvas?.getContext('2d');

  ctx?.clearRect(0, 0, 300, 100);
  state.signature = '';
}

watch(state, () => {
  handleCanvasUpdate();
});

function handleCanvasUpdate() {
  const canvas = signatureCanvas.value;
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //@ts-ignore
  const ctx = canvas.getContext('2d');

  if (ctx) {
    ctx.font = '30px Brush Script MT';
    ctx.clearRect(0, 0, 300, 100);
    ctx.fillText(state.signature, 10, 50);
  }
}
</script>

<style scoped lang="scss">
.signature-preview {
  border: 1px solid #333;
  border-radius: 10px;
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  align-items: flex-end;
  margin-left: 1rem;
}
</style>
