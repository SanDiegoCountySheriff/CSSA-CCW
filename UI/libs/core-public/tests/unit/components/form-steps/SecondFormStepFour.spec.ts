import SecondFormStepFour from '@core-public/components/form-stepper/form-steps/SecondFormStepFour.vue';
import Vuetify from 'vuetify';
import { createLocalVue, mount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';

const pinia = createTestingPinia();
const localVue = createLocalVue();
const tMock = {
  $t: text => text,
};

describe('SecondFormStepFour', () => {
  let vuetify;
  let wrapper;
  beforeEach(() => {
    vuetify = new Vuetify();
    wrapper = mount(SecondFormStepFour, {
      localVue,
      vuetify,
      pinia,
      mocks: tMock,
      propsData: {
        handleNextSerction: () => null,
      },
    });
  });
  it('Should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
