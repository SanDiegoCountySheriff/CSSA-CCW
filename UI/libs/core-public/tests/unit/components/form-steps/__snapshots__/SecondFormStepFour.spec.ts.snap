// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecondFormStepFour Should match the snapshot 1`] = `
<div>
  <div class="container">
    <div class="row mb-5">
      <div class="col-sm-4 col-md-5 col-10">
        <form novalidate="novalidate" class="v-form">
          <div class="v-input theme--light v-text-field">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-3" class="v-label theme--light" style="left: 0px; position: absolute;">Signature</label><input id="input-3" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
      <div class="col-sm-4 col-md-5 col-10">
        <div class="signature-preview"><canvas height="100" width="300"></canvas> <button type="button" class="m-5 v-btn v-btn--text theme--light v-size--default error--text"><span class="v-btn__content">
            clear
          </span></button></div>
      </div>
    </div>
    <hr role="separator" aria-orientation="horizontal" class="mb-5 v-divider theme--light">
    <div class="form-btn-container">
      <div><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default"><span class="v-btn__content">
      Continue
    </span></button> <button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default secondary mx-2"><span class="v-btn__content">
      Save and Exit
    </span></button></div>
      <div><button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default warning mx-2"><span class="v-btn__content">
      Go back
    </span></button> <button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default error mr-2"><span class="v-btn__content">
      Cancel
    </span></button></div>
    </div>
  </div>
  <div class="v-snack v-snack--bottom">
    <transition-stub name="v-snack-transition">
      <div class="v-snack__wrapper v-sheet v-sheet--outlined theme--light error--text" style="display: none;">
        <div role="status" aria-live="polite" class="v-snack__content">
          File upload unsuccessful please try again.
        </div>
        <div class="v-snack__action "></div>
      </div>
    </transition-stub>
  </div>
</div>
`;
